### PR TITLE
Beam 12994 - Python SDK BigQuery - Promote `schemaUpdateOptions` to named arguments

### PR DESCRIPTION
**Pull request by waltage**
### **Python - IO - GCP - BigQuery:**
Promotes **_schema update options_** that are currently available as entries in the `additional_bq_parameters` dict to named and verified arguments (similar to the current Create and Write Dispositions).

addresses #21141 
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by waltage** [comment]
following up on the status of this review

**Comment by waltage** [comment]
@aaltay @pabloem 
I recognize this is a lower priority issue/PR, but it's my first PR with Apache and on Beam.  I'd like to move on to some higher priority issues for you all, but this review is currently a blocker

**Comment by johnjcasey** [comment]
This looks structurally fine to me. My only question is if we have an integration test suite that could be expanded to include a test that leverages these parameter, to verify they work as expected

**Comment by waltage** [comment]
> This looks structurally fine to me. My only question is if we have an integration test suite that could be expanded to include a test that leverages these parameter, to verify they work as expected

If you could point me in the right direction (like a particular directory or a test file), I can work on adding one.  

My original thinking was that any integration testing on this particular "feature" would reduce to a change-detector test given that the implementation is setting the parameter on the proto request itself ([here](https://github.com/apache/beam/pull/21867/commits/4083fa49cb6e9955b2f7eca70727c8571b5652ed#diff-5427a5d3887eb695cefde082c58575a2372972996b547d55961abeb4f7bc3debR534)), but it is duly noted that the string values themselves are not explicitly being tested anywhere

**Comment by waltage** [comment]
@pabloem no worries on the review timeline.

I am not the original reporter of the Jira issue, but it looks like this particular feature was requested with an objective of bringing the Python API closer in line with Java API.  Since the issue was ported over to github and tagged with first issue, and since Ken in the original issue said it seemed like a good idea, I went ahead and worked on implementing it here.

I agree that this particular initializer is quite dense, regardless of whether this option is promoted to a parameter.  I think there are a few ways I could go with this:
1. (assuming I can update the other unit tests that use this transform, and provide  sufficient integration coverage) Allow this option promotion in light of the already dense parameter set.
2. Develop type and value checking for the ...bq_parameters... dict, which would allow the surface of the initializer to remain consistent.
3. Close the issue/pr, as interest looks like it has subsided.
4. Abstract away a **majority** of the transform's initializer parameters that are BigQuery-specific into a new type/value-checked Class that would allow more flexibility for changes like this (or changes to the BigQuery APIs themselves) without modifying existing surfaces of the transform itself.

In a related note, as a tie-breaker for 1 & 2 & 3, and in support of the design of 4 above, does the BEAM team currently analyze any of the public usages of its python libraries?  I can look at how this transform is used "in the wild" and try to understand how often parameters are used vs. what keys of the dict are typically set (e.g., at first glance, it looks like partitioning specs are by far the most commonly used keys in the ...bq_parameters... dict).

I'm happy to work on any of the above, just let me know how best to proceed.

**Comment by pabloem** [comment]
thanks for the explanation @waltage - I am going to be on vacation for a little while, so I can't advice at the moment.

@ahmedabu98 @Abacn @johnjcasey could y'all help Daniel figure out how to move forward? thanks

**Comment by johnjcasey** [comment]
I am ok with the dense parameter set. The Beam convention is to use named parameters in python like we use builder parameters in Java. If something has a default, that default must be valid such that the user doesn't need to provide it.

I like brining python in line with Java, and I prefer having a named parameter to a dict for additional parameters

File: sdks/python/apache_beam/io/gcp/bigquery_tools.py [hunk 1]
```diff
@@ -527,6 +531,7 @@ def _insert_load_job(
                     schema=job_schema,
                     writeDisposition=write_disposition,
                     createDisposition=create_disposition,
+                    schemaUpdateOptions=schema_update_options,
                     sourceFormat=source_format,
                     useAvroLogicalTypes=True,
                     autodetect=schema == 'SCHEMA_AUTODETECT',
```
Comment (ahmedabu98) [inline sdks/python/apache_beam/io/gcp/bigquery_tools.py:537]: We should confirm early in WriteToBigQuery that the schema update options are present in either `schema_update_options` or `additional_load_parameters` and resolve if options are present in both. 

If options are present in `additional_load_parameters` at this point, we would have two `schemaUpdateOptions` parameters here.

Comment (ahmedabu98) [reply sdks/python/apache_beam/io/gcp/bigquery_tools.py:537]: We should also allow backwards compatibility. If users want to keep using `additional_load_parameters` instead to specify update options, would they be able to?

**Review by ahmedabu98 — CHANGES_REQUESTED** [review]
Thanks for working on this @waltage! Left a couple of comments.

I can't speak on the user demand of having this option as a parameter in WriteToBigQuery, but I don't see a problem with adding it as long as users can still specify these options in `additional_load_parameters`.

**Comment by ahmedabu98** [comment]
Agree with @johnjcasey that we need an integration test for this change, specifically to make sure that both `additional_load_parameters` and `schema_update_options` are viable. I would be happy to review that test.

@waltage we have a few BQ tests in [apache_beam/io/gcp](https://github.com/apache/beam/tree/master/sdks/python/apache_beam/io/gcp). This is an [example test](https://github.com/apache/beam/blob/82384a462a04c046723d8dcd8377373ecc5767d4/sdks/python/apache_beam/io/gcp/bigquery_test.py#L1774-L1842) that uses [`BigQueryTableMatcher`](https://github.com/apache/beam/blob/c36f54486a5185fd11bc34186a688e9b7abfa4b5/sdks/python/apache_beam/io/gcp/tests/bigquery_matcher.py#L207) to test for options in `additional_bq_parameters`.

**Comment by ahmedabu98** [comment]
FYI would need to see what existing checks are performed against `additional_parameters["schemaUpdateOptions"]` and make sure the new `schemaUpdateOptions` argument is there too. e.g. [here](https://github.com/apache/beam/blob/7a9bb76fe9f4c167c1d125db9d2cff9a1a315149/sdks/python/apache_beam/io/gcp/bigquery_file_loads.py#L380-L384).

**Pull request closed without merging**